### DISCUSSION TITLE
Method for list of all model files

### DISCUSSION THREAD

**Discussion by laat0003** [modflowpy/flopy#1637] [Q&A] [answered]
Hi,

Is there a method to obtain a list of all model files required to execute a simulation? For example, I have a simulation that includes parent and child models, both with external arrays, time series files, observation files, advanced stress package input files, exchange files plus the standard package input files. When writing the simulation flopy is aware of all the filenames used in the simulation so I assume it must be accessible, but I am too stoopid :)

Cheers,
Tariq

**Comment by spaulins-usgs** [comment, marked as answer]
@laat0003, there is unfortunately no single method that returns a list of all files in a simulation.  However, by iterating through a tree of objects you should be able to do this.  I can add a method to flopy that does this, but I may not get to it for a little while.  If you would like to try to do this yourself, below I describe the tree structure you will need to traverse to get the filenames.  The objects are in **bold** and the methods/attributes used to access those objects are in _italics_.

* **Simulation**
  * _name_file_
    * **Namefile Package**
  * _get_package(None)_ 
    * **Package1** 
    * **Package2** 
  * _model_dict_
     * **Model1** 
     * **Model2** 
       * _name_file_
         * **Namefile Package** 
       * _get_package(None)_
         * **Package1**
         * **Package2**
           * _filename_
             * **Package's filename string** 
           * _blocks_
             * **Block1**
             * **Block2**
               * _datasets_  
                 * **Dataset1**
                 * **Dataset2** 
                   * _get_record_
                      **dictionary tree of data and metadata** (for any external data there will be a "filename" key in the lowest level dictionaries)


At the top level, the simulation object can contain both model and package objects.  Use the _model_dict_ property to get all of the simulation's model objects in a dictionary.   You can get all the simulation's package objects by calling the _get_package_ method (without parameters).  Once you have model and package objects, you can get any packages in the model or sub-packages in the package with the same _get_package_ call.  Use the _filename_ property to access each package's filename.

In order to get the files of external datasets you will need to iterate further down the tree into the package's blocks and data.  The package's _blocks_ attribute is a dictionary of the package's blocks.  Similarly, each block has a _datasets_ property.  Dataset's that can be stored externally have a _get_record_ method that returns a dictionary.  Note that datasets that can not be stored externally do not have the _get_record_ method, so you will need to check for this "hasattr(obj, 'get_record')".

Since some datasets can be divided up by layer and by stress period and each part of the dataset can be stored separately, the dictionary returned may be a dictionary of dictionaries, with the first dictionary pointing to each layer or stress period of the data.  Iterate through the dictionary tree searching for "filename" keys, whose value will be the filename of the stored data.

Lastly, the simulation and model name files will not be returned with the other packages.  This in intentional since, in general, it is a bad idea to mess with these.  Flopy builds these and keeps them updated for you.  The name files can be accessed from the simulation and model objects with the _name_file_ attribute.  The name file objects have the same _filename_ attribute as the other packages.

**Reply by laat0003** [reply]
Thanks for the very comprehensive answer @spaulins-usgs. It is greatly appreciated. I had tried some of what you mention already but wasn't getting all the files. The tree structure you provided is perfect for me to work off.  You guys rock. 

All the best.
